cdk/lib/auth/pretokengeneration/index.ts

```diff
@@ -0,0 +1,51 @@
+import AWS = require("aws-sdk");
+import { ListUsersRequest } from "aws-sdk/clients/cognitoidentityserviceprovider";
+
+export const handler = async (event: any) => {
+  // email of user who just signed in
+  const email = event.request.userAttributes["email"];
+
+  if (!email || !process.env.USER_POOL_ID) {
+    return event;
+  }
+
+  var cognitoidentityserviceprovider = new AWS.CognitoIdentityServiceProvider();
+
+  // search in Cognito for user with that email
+  var params: ListUsersRequest = {
+    UserPoolId: process.env.USER_POOL_ID,
+    Filter: `email=\"${email}\"`,
+  };
+  // may get back multiple as Cognito will create a copy of a user the IdP
+  var users = await cognitoidentityserviceprovider.listUsers(params).promise();
+
+  var roles: string = '[""]';
+
+  if (users && Array.isArray(users["Users"])) {
+    // filter out copy of user in IdP
+    var user = users["Users"].filter(
+      (attr) => attr["UserStatus"] != "EXTERNAL_PROVIDER"
+    );
+
+    if (user.length == 1 && Array.isArray(user[0]["Attributes"])) {
+      // filter out everything but the roles attribute
+      var roleAttribute = user[0]["Attributes"].filter(
+        (attr) => attr["Name"] == "custom:roles"
+      );
+      if (roleAttribute.length == 1 && roleAttribute[0]["Value"])
+        roles = roleAttribute[0]["Value"];
+    }
+  }
+
+  // override the claim with the roles in Cognito for the user who's signing in
+  event.response = {
+    claimsOverrideDetails: {
+      claimsToAddOrOverride: {
+        "custom:roles": roles,
+      },
+    },
+  };
+  console.log(event.response);
+
+  return event;
+};
```

Comment:
I'm worried about the complexity that we are adding to our Auth workflow here. I know we have discussed this before but we should consider offloading this complexity to Cognito. What I mean is that we shouldn't create a record for a federated user in our Cognito User Pool. Let Cognito manage those records for external identities. Having to maintain this "user clone" is opening the door for issues.

Reply:
The more agnostic our application can be to Federation, the better. Let customers configure federation in Cognito if they want, but our application shouldn't need to change or know if the users are coming from an external identity provider or from Cognito.

Reply:
@ferdingler That's what I recommended initially, that we don't manage users in Badger at all.  But since we create users and roles in Badger, then we have to map the federated user to the Badger user.  I believe Quicksights does the same

Reply:
That's correct, customers can configure federation in Cognito, and it wouldn't affect Badger.  But once they set EnvConfig to point Badger to use federation, then Badger must know to do a federated sign in, which is a different flow